lib/Sema/TypeCheckPattern.cpp

```diff
@@ -1007,6 +1005,10 @@ bool TypeChecker::coercePatternToType(Pattern *&P, TypeResolution resolution,
       // Assume the compiler knows what it's doing.
     } else if (diagTy->isEqual(Context.TheEmptyTupleType)) {
       shouldRequireType = true;
+    } else if (auto arrTy = dyn_cast<ArraySliceType>(diagTy.getPointer())) {
```

Comment:
I don't think you want to check for this specifically: that's looking for something that's *syntactically* using brackets, which means it'll ignore `Array<T>` if someone decides to write it.

Reply:
My bad, I originally put this suggestion in the SR. I think you can use `getArrayDecl()` (from ASTContext)  for comparison (similar to how `ConstraintSystem::isArrayType`) works. I don't think we have a method that checks for both `[T]` and `Array<T>`, might be worth defining one.

Reply:
`[T]` is sugar for `Array<T>` anyway, so checking against `getArrayDecl()` is the way to go. (See also #27195 for a potential simplification in this area.)

Reply:
Great, thanks for catching this. I am happy to wait for (or help with) #27195 and then update here.